pkg/mcp/server/server.go

```diff
@@ -99,11 +125,84 @@ type AuthChecker interface {
 	Check(authInfo credentials.AuthInfo) error
 }
 
+type mailboxState int
+
+const (
+	mailboxStateReady mailboxState = iota
+	mailboxStateClosed
+)
+
 // watch maintains local state of the most recent watch per-type.
 type watch struct {
-	cancel       func()
-	nonce        string
-	responseChan chan *WatchResponse
+	// only accessed from connection goroutine
+	cancel          func()
+	nonce           string // most recent nonce
+	nonceVersionMap map[string]string
+
+	// NOTE: do not hold `mu` when reading/writing to this channel
+	mailboxReadyChan chan mailboxState
+
+	mu                      sync.Mutex
+	mailbox                 *WatchResponse
```

Comment:
Perhaps a more descriptive name than mailbox? It looks like this should be really the current response, or something similar.

Would it make sense to create a separate primitive (i.e. a struct) that captures mailboxReadyChan & mailbox (and maybe mu) for better factoring?

Reply:
newPushResponse?

The current fields that need to be protected are grouped under `mu`, which is a common idiom. And we _don't_ want to hold the lock for newPushresponseReadyChan. It's not clear whether we should group (via struct) by what needs to be protected by the lock (newPushResponse, mostRecentNackedVersion, timer) or by function (newPushResponse, newPushResponseReadyChan), or both, in which case keeping all of the watch push state in a single struct seems simpler.